[CAS-1425] Events handler iteration (#2588)
GetStream/stream-chat-android: docusaurus/docs/Android/03-ui/04-components/02-channel-list.mdx

        ),
        sort = ChannelListViewModel.DEFAULT_SORT,
        limit = 30,
        chatEventHandler = DefaultChatEventHandler(),
    )
}
// Bind the ViewModel with ChannelListView
viewModel.bindView(channelListView, viewLifecycleOwner)
```

> Here you see to instantiate `ChannelListViewModel` you need to pass an instance of the `ChatEventHandler` interface. See [this point](##ChatEventHandler) to read more.

Reviewer: ```suggestion
You need to pass an instance of the `ChatEventHandler` interface in order to instantiate `ChannelListViewModel`. See [this point](##ChatEventHandler) for further details.
```

These changes should have the following results:

![Light mode](../../assets/channel_list_shimmer.png)

## ChatEventHandler

> When `ChannelListViewModel` is instantiated it makes a query channels request. It gets data and shows in the `ChannelListView`. With time the list gets outdated and to keep it sync we have websocket events. But when handling events locally, our SDK can't always decide what we should do with the list. For example, in a messaging app it's common to add a channel to the list if the current user became a channel member and to remove this channel otherwise. But the same cannot be applied in the use case of the livestream app. To define desired behaviour there is `ChatEventHandler` which has a following interface:

Reviewer: ```suggestion
When `ChannelListViewModel` is instantiated it makes a query channels request to get the list of channels. Starting from this point, the list is being synced through WebSocket events. The SDK includes a `ChatEventHandler` that decides whether to include or exclude the channel from the list after receiving some events. Sometimes desired behavior differs from the default one and you can notice that for example - the channel doesn't disappear from the list after removing a member. In that case, you should provide a custom `ChatEventHandler` implementation, which has the following interface:
```

```kotlin
public fun interface ChatEventHandler {
    /**
     * Function that computes result of handling event. It runs in background.
     *
     * @param event ChatEvent that may contain updates for the set of channels. See more [ChatEvent]
     * @param filter [FilterObject] that can be used to define result of handling.
     *
     * @return [EventHandlingResult] Result of handling.
     */
    public fun onChatEvent(event: ChatEvent, filter: FilterObject): EventHandlingResult
}

public enum class EventHandlingResult {
    /**
     * Add a channel to a query channels collection.
     */
    ADD,

    /**
     * Remove a channel from a query channels collection.
     */
    REMOVE,

    /**
     * Skip handling of this event.
     */
    SKIP
}
```

> We also provide the `BaseChatEventHandler` implementation. It's an abstract incomplete implementation that covers the most generic cases and forces you to implement the required ones.

Reviewer: ```suggestion
You can also use `BaseChatEventHandler` - a more detailed implementation that covers the most generic cases and forces you to implement the required ones.
```

These changes should have the following results:

![Light mode](../../assets/channel_list_shimmer.png)

## ChatEventHandler

> When `ChannelListViewModel` is instantiated it makes a query channels request. It gets data and shows in the `ChannelListView`. With time the list gets outdated and to keep it sync we have websocket events. But when handle event our SDK locally can't always decide what we should do with the list. For example, in a messaging app it's common to add a channel to the list if the current user became a channel member and to remove this channel otherwise. But the same cannot be applied in the use case of the livestream app. To define desired behaviour there is `ChatEventHandler` which has a following interface:

Reviewer 2: `when handle event our SDK locally` -> `when handling events locally, our SDK`

```kotlin
public fun interface ChatEventHandler {
    /**
     * Function that computes result of handling event. It runs in background.
     *
     * @param event ChatEvent that may contain updates for the set of channels. See more [ChatEvent]
     * @param filter [FilterObject] that can be used to define result of handling.
     *
     * @return [EventHandlingResult] Result of handling.
     */
    public fun onChatEvent(event: ChatEvent, filter: FilterObject): EventHandlingResult
}

public enum class EventHandlingResult {
    /**
     * Add a channel to a query channels collection.
     */
    ADD,

    /**
     * Remove a channel from a query channels collection.
     */
    REMOVE,

    /**
     * Skip handling of this event.
     */
    SKIP
}
```

We also provide the following implementation of this interface:

1. `MessagingChatEventHandler`. It is based on assumption when the current user becomes a member of a channel this channel should be added to the channel list, otherwise deleted.

> 2. `DefaultChatEventHandler`. More generic version, it skips events with update payload and makes an api request everytime when receives `NotificationAddedToChannelEvent`, `NotificationMessageNewEvent`, `NotificationRemovedFromChannelEvent` events. It can cause additional expenses on your plan, be sure when use this implementation.

Reviewer 2: `everytime when receives`-> `every time it receives`

`be sure when use this implementation` -> `make sure you understand when to use this implementation` or something @zsmb13 maybe can have a better suggestion.

Is there a reason why the Default handler is worse for the user and the MessaggingChatEventHandler is not the default?

Author: We removed it. So it's not actual anymore

> 3. `BaseChatEventHandler`. It's abstract not complete implementation that covers most generic cases and force to implement you required ones. You can inherit it and define behaviour for your custom use-case.

Reviewer 2: `it's an  abstract incomplete implementation that covers the most generic cases and forces you to implement the required ones.`
